src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs

```diff
@@ -0,0 +1,358 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PowerFx.Core.Utils;
+
+namespace Microsoft.PowerFx.Core.Functions
+{
+    internal class TexlFunctionSet<T>
+        where T : ITexlFunction
+    {
+        private Dictionary<string, List<T>> _functions;
+        private Dictionary<string, List<T>> _functionsInvariant;
+        private Dictionary<DPath, List<T>> _namespaces;
+        private List<string> _enums;
+        private int _count;
+
+        internal TexlFunctionSet()
+        {
+            _functions = new Dictionary<string, List<T>>();
+            _functionsInvariant = new Dictionary<string, List<T>>();
+            _namespaces = new Dictionary<DPath, List<T>>();
+            _enums = new List<string>();
+            _count = 0;
+        }
+
+        internal TexlFunctionSet(T function)
+            : this()
+        {
+            if (function == null)
+            {
+                throw new ArgumentNullException($"{nameof(function)} cannot be null", nameof(function));
+            }
+
+            _functions.Add(function.Name, new List<T>() { function });
+            _functionsInvariant.Add(function.LocaleInvariantName, new List<T>() { function });
+            _namespaces.Add(function.Namespace, new List<T> { function });
+            _enums = function.GetRequiredEnumNames().ToList();
+            _count = 1;
+        }
+
+        internal TexlFunctionSet(IEnumerable<T> functions)
+            : this()
+        {
+            if (functions == null)
+            {
+                throw new ArgumentNullException($"{nameof(functions)} cannot be null", nameof(functions));
```

Comment:
>$"{nameof(functions)} cannot be null", [](http://example.com/codeflow?start=48&length=38)

super nit - this arg isn't necessary. ArgumentNullException will make the message from just a single arg

Reply:
this is for making sure we'll get a localized message when we'll have to deal with this later

Reply:
Do we intend to localize exception messages? That can make reviewing telemetry more complicated than needed. generally, the exception messages should be dev-facing, not user-facing.

Reply:
For telemetry we should have an ErrorId

Reply:
fixed it